labels: Flatten Labels and LabelArray

jrajahalme opened the pull request:
Labels are typically immutable after creation, except for the
field 'deletionMark'. As deletionMark is only used in conjunction with
OpLabels, refactor the code to use the marking functionality within OpLabels
instead of having the mutable deletionMark field in the Label itself.

Store Labels by value rather than pointers as this will generally
be faster due to less ping-ponging across cachelines.

Removing the deletionMark from the Label and storing the labels by
value removes the need for DeepCopy()ing Labels and simplifies the
DeepCopy implementation for LabelArrays.

These changes make manipulations as searches on a synthetic test that
creates 10000 rules about 30% faster and 10% more memory efficient.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6014)
<!-- Reviewable:end -->

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -2239,7 +2200,7 @@ func APICanModify(e *Endpoint) error {
 		return nil
 	}
-	if lbls := e.OpLabels.OrchestrationIdentity.FindReserved(); lbls != nil {
+	if e.OpLabels.IsReserved() {
```
Leave the `OrchestrationIdentity` with the `IsReserved()` it's misleading to check if the opLabels are reserved but only doing the check for the `OrchestrationIdentity` labels.

jrajahalme replied:
OK, will do.

jrajahalme replied:
I made `IsReserved()` a method of `Labels` that returns `true` if any of the labels has a reserved source. Runs a bit faster without creating a map.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -2268,63 +2229,39 @@ func (e *Endpoint) ModifyIdentityLabels(owner Owner, addLabels, delLabels pkgLab
 
 	switch e.GetStateLocked() {
 	case StateDisconnected, StateDisconnecting:
+		e.Unlock() // XXX: Bug Fix
```
what bug fix? leave more comments

jrajahalme replied:
I'll separate this to another commit with it's own commit message to make this clear. This case was missing unlock and we were returning from the function with endpoint's lock held.

jrajahalme replied:
Done (#6145)

aanm commented on pkg/labels/array.go:
```diff
@@ -135,23 +135,29 @@ func (ls LabelArray) DeepCopy() LabelArray {
 }
 
 // GetModel returns the LabelArray as a string array with fully-qualified labels.
 // The output is parseable by ParseLabelArrayFromArray
```
Is this command still valid?

jrajahalme replied:
Should be.

aanm replied:
Sorry, I meant comment/description. Is the output still parseable by `ParseLabelArrayFromArray`?

aanm replied:
@jrajahalme missing

jrajahalme replied:
Yes as the output format is not changed.

aanm commented on pkg/labels/array.go:
```diff
@@ -135,23 +135,29 @@ func (ls LabelArray) DeepCopy() LabelArray {
+	res := "["
+	for l := range ls {
+		if l > 0 {
+			res += " "
```
why not `", "`?

jrajahalme replied:
I made the output here match the one we got previously from a slice of pointers. If we add the comma we also need to adjust a bunch of test cases.

aanm commented on pkg/labels/zz_generated.deepcopy.go:
```diff
@@ -1,82 +0,0 @@
-// +build !ignore_autogenerated
```
I might be wrong but I think we need a deepcopy function for kubernetes, try doing `make generate-k8s-api`

jrajahalme replied:
We don't have references to any of the generated functions anywhere in the repo anymore, so I don't see how they would be needed? I did not find how this file was gnerated, however. Wuld this be re-generated by `make generate-k8s-api`?

jrajahalme replied:
Had to `export GOPATH=$HOME/go` before `make generate-k8s-api` would run in the dev VM, but after that it runs fine and re-creates `pkg/labels/zz_generated.deepcopy.go`. If I delete `pkg/labels/zz_generated.deepcopy.go` and `make` it builds, so the file is not needed.

Figured out how to edit the Makefile to not generate deepcopy functions for pkg/labels.

aanm replied:
@jrajahalme remove the `pkg/labels/zz_generated.deepcopy.go` and apply the following diff.
```
diff --git a/pkg/labels/doc.go b/pkg/labels/doc.go
index db47aa6b8..7931b1922 100644
--- a/pkg/labels/doc.go
+++ b/pkg/labels/doc.go
@@ -12,9 +12,5 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// +k8s:deepcopy-gen=package
-// +k8s:openapi-gen=true
-
 // Package api defines the API of the Cilium network policy interface
-// +groupName=pkg
 package labels
```

aanm replied:
@jrajahalme missing?

jrajahalme replied:
Done!

jrajahalme commented on pkg/labels/array.go:
```diff
@@ -91,17 +91,17 @@ nextLabel:
 		for _, lsl := range ls {
-			if lsl.GetExtendedKey() == key {
+			// Note that if '=value' is part of 'key' it is ignored here
+			if lsl.Source == keyLabel.Source && lsl.Key == keyLabel.Key {
```
@aanm Please have a look here. According to the comment above the key is of from `source.key=value`. If the `=value` part is really in there, then we used to never match that as `GetExtendedKey()` does NOT include the value in the returned string. It may be that the comment is wrong and the key never has the `=value` part, though.

aanm replied:
@jrajahalme where is this code now?

aanm replied:
@jrajahalme missing?

jrajahalme replied:
This was merged already in #5967 that I had used as the basis for this PR. Please have look: https://github.com/cilium/cilium/pull/5967/files.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -2268,6 +2268,7 @@ func (e *Endpoint) ModifyIdentityLabels(owner Owner, addLabels, delLabels pkgLab
 
 	switch e.GetStateLocked() {
 	case StateDisconnected, StateDisconnecting:
+		e.Unlock()
```
Can you push this into a separate PR? This warrants backporting.

jrajahalme replied:
Done (#6145)

ianvernon commented on pkg/labels/oplabels.go:
```diff
@@ -0,0 +1,246 @@
+	// Labels derived from orchestration system
+	OrchestrationIdentity Labels
+
+	//orchestrationIdentity
```
Remove this line

jrajahalme replied:
Done.

ianvernon commented on pkg/labels/oplabels.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+// GetIdentityLabel returns the value of the given Key from all IdentityLabels.
+func (o *OpLabels) GetIdentityLabel(key string) (l Label, found bool) {
```
Shouldn't we return both labels from `Custom` and `OrchestrationIdentity` if they map to different values? I'm not sure if such a case is even allowed, though.

jrajahalme replied:
@aanm may correct me if I am wrong, but IMO we only ever have a given key in one of the sets.

aanm replied:
I think that's is correct @jrajahalme

ianvernon commented on pkg/labels/oplabels.go:
```diff
@@ -0,0 +1,246 @@
+	return changed
+}
+
+func (o *OpLabels) ModifyIdentityLabels(addLabels, delLabels Labels) (changed bool, err error) {
```
Can you file a GH issue to add unit tests for these functions?

jrajahalme replied:
Issued #6146

ianvernon commented on pkg/labels/oplabels.go:
```diff
@@ -0,0 +1,246 @@
+		if _, found := o.Disabled[k]; found {
+			disabledKeepers.set(k)
+		} else if keepers.set(v.Key); o.OrchestrationIdentity.upsertLabel(v) {
+			if logger != nil {
```
Is logger not initialized as part of the package being imported? This check shouldn't be needed.

jrajahalme replied:
`logger` here is the endpoint's logger passed in as an argument. We do call `UpdateLogger()` on all code paths that create endpoints. though, so it will not be nil. Changed accordingly.

aanm commented on pkg/labels/oplabels.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+// OpLabels represents the the possible types.
+// +k8s:openapi-gen=false
```
Remove these 2 lines

jrajahalme replied:
Done.

ianvernon reviewed (approved):
Approving based off of prior review + since you addressed all of my comments directly. Thanks!

jrajahalme committed:
labels: Flatten Labels and LabelArray

Labels are typically immutable after creation, except for the
field 'deletionMark'. As deletionMark is only used in conjunction with
OpLabels, refactor the code so that the deletion marks can be kept
local to the functions that need them.

Store Labels by value rather than pointers as this will generally
be faster due to less ping-ponging accross cachelines.

Removing the deletionMark from the Label and storing the labels by
value removes the need for DeepCopy()ing Labels and simplifies the
DeepCopy implementation for LabelArrays.

These changes make manimulations as searches on a synthetic test that
creates 10000 rules about 30% faster and 10% more memory efficient.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Repository benchmark.
